File: 백준/Silver/9012. 괄호/괄호.c
#include <stdio.h>
#include <string.h>
#define STACK 51

int main() {
	int T;
	char stack[STACK];
	scanf("%d",&T);
	for(int i = 0; i < T; i++){
		int j=0,sum = 0;
		scanf("%s", stack);
		while(stack[j]!=NULL) {
				if (stack[j] == '(') {
					sum++;
				}
				if (stack[j] == ')') {
					sum--;
				}
				if (sum < 0)
				{
					printf("NO\n");
					break;
				}
				j++;
		}
		if (sum == 0) {
			printf("YES\n");
		}
		else if(sum>0) {
			printf("NO\n");
		}
	}
	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.py
t = int(input())

for _ in range(t):
    ps = input()
    stack = []
    
    for p in ps:
        if p == '(':
            stack.append('(')
        elif p == ')':
            if len(stack) == 0:
                stack.append(')')
                break
            else:
                stack.pop()
        
    if len(stack) != 0:
        print('NO')
    else:
        print('YES')